src/Common/AbstractEnum.php

```diff
@@ -0,0 +1,356 @@
+<?php
+
+declare(strict_types=1);
+
+namespace WordPress\AiClient\Common;
+
+use BadMethodCallException;
+use InvalidArgumentException;
+use ReflectionClass;
+
+/**
+ * Abstract base class for enum-like behavior in PHP 7.4
+ *
+ * This class provides enum-like functionality for PHP versions that don't support native enums.
+ * Child classes should define uppercase snake_case constants for enum values.
+ *
+ * @example
+ * class PersonEnum extends AbstractEnum {
+ *     public const FIRST_NAME = 'first';
+ *     public const LAST_NAME = 'last';
+ * }
+ *
+ * // Usage:
+ * $enum = PersonEnum::from('first'); // Creates instance with value 'first'
+ * $enum = PersonEnum::tryFrom('invalid'); // Returns null
+ * $enum = PersonEnum::firstName(); // Creates instance with value 'first'
+ * $enum->name; // 'FIRST_NAME'
+ * $enum->value; // 'first'
+ * $enum->equals('first'); // Returns true
+ * $enum->is(PersonEnum::firstName()); // Returns true
+ * PersonEnum::cases(); // Returns array of all enum instances
+ */
+abstract class AbstractEnum
+{
+    /**
+     * @var string|int The value of the enum instance
+     */
+    private $value;
+
+    /**
+     * @var string The name of the enum constant
+     */
+    private $name;
+
+    /**
+     * @var array<string, array<string, string|int>> Cache for reflection data
+     */
+    private static $cache = [];
+
+    /**
+     * @var array<string, array<string, self>> Cache for enum instances
+     */
+    private static $instances = [];
+
+    /**
+     * Constructor is private to ensure instances are created through static methods
+     *
+     * @param string|int $value The enum value
+     * @param string $name The constant name
+     */
+    private function __construct($value, string $name)
+    {
+        $this->value = $value;
+        $this->name = $name;
+    }
+
+    /**
+     * Magic getter to provide read-only access to properties
+     *
+     * @param string $property The property name
+     * @return mixed
+     * @throws BadMethodCallException If property doesn't exist
+     */
+    public function __get(string $property)
```

Comment:
Since none of the methods in this class should ever be overwritten, but the class _does_ need to be abstract, I think it would help to make all public and protected methods in this class `final`, to clarify that interface and make sure developers can't mess with it in unexpected ways.

Reply:
Agreed! Resolved in 71529cb